src/main/java/org/rumbledb/compiler/TranslationVisitor.java

```diff
@@ -234,111 +232,110 @@ public Node visitExprSingle(JsoniqParser.ExprSingleContext ctx) {
             } else if (child instanceof JsoniqParser.WhereClauseContext) {
-                childClause = (Clause) this.visitWhereClause((JsoniqParser.WhereClauseContext) child);
+                clause = (Clause) this.visitWhereClause((JsoniqParser.WhereClauseContext) child);
             } else if (child instanceof JsoniqParser.GroupByClauseContext) {
-                childClause = (Clause) this.visitGroupByClause((JsoniqParser.GroupByClauseContext) child);
+                clause = (Clause) this.visitGroupByClause((JsoniqParser.GroupByClauseContext) child);
             } else if (child instanceof JsoniqParser.OrderByClauseContext) {
-                childClause = (Clause) this.visitOrderByClause((JsoniqParser.OrderByClauseContext) child);
+                clause = (Clause) this.visitOrderByClause((JsoniqParser.OrderByClauseContext) child);
             } else if (child instanceof JsoniqParser.CountClauseContext) {
-                childClause = (Clause) this.visitCountClause((JsoniqParser.CountClauseContext) child);
+                clause = (Clause) this.visitCountClause((JsoniqParser.CountClauseContext) child);
             } else {
                 throw new UnsupportedFeatureException(
                         "FLOWR clause not implemented yet",
                         createMetadataFromContext(ctx)
                 );
             }
-            childClause.setPreviousClause(previousFLWORClause);
-            previousFLWORClause = childClause;
-            contentClauses.add(childClause);
+
+            previousFLWORClause.chainWith(clause.getFirstClause());
+            previousFLWORClause = clause.getLastClause();
         }
 
         // visit return
 
         Expression returnExpr = (Expression) this.visitExprSingle(ctx.return_expr);
-        returnClause = new ReturnClause(
+        ReturnClause returnClause = new ReturnClause(
                 returnExpr,
                 new ExceptionMetadata(
                         ctx.getStop().getLine(),
                         ctx.getStop().getCharPositionInLine()
                 )
         );
-        returnClause.setPreviousClause(previousFLWORClause);
+        previousFLWORClause.chainWith(returnClause);
 
         return new FlworExpression(
-                startClause,
-                contentClauses,
                 returnClause,
                 createMetadataFromContext(ctx)
         );
     }
 
     @Override
     public Node visitForClause(JsoniqParser.ForClauseContext ctx) {
-        List<ForClauseVar> vars = new ArrayList<>();
-        ForClauseVar child;
+        ForClause clause = null;
         for (JsoniqParser.ForVarContext var : ctx.vars) {
-            child = (ForClauseVar) this.visitForVar(var);
-            vars.add(child);
+            ForClause newClause = (ForClause) this.visitForVar(var);
+            if (clause != null) {
+                clause.chainWith(newClause.getFirstClause());
+            }
+            clause = (ForClause) newClause.getLastClause();
```

Comment:
Are the getFirstClause and getLastClause calls needed here? Isn't the result of visitForVar a single clause that is non-chained? I see that getFirstClause and getLastClause calls are idempotent when the clause chain consists of only 1 function so this is a non-issue from a correctness perspective. Yet, I was curious if there is some underlying chaining with variable definitions that I am missing.

Reply:
Very good catch! I simplified.